samza-yarn/src/main/java/org/apache/samza/util/hadoop/CertFSFileSystem.java

```diff
@@ -0,0 +1,461 @@
+   *
+   * @return <code>certfs</code>
+   */
+  @Override
+  public String getScheme() {
+    return _uri.getScheme();
+  }
+
+  @Override
+  public URI getUri() {
+    return _uri;
+  }
+
+  /**
+   * convert certfs URI to https URI
+   * @param certFsUri certFS URI where the CSR should be sent to
+   * @param excludesQuery whether or not to exclude the query params
+   * @return https URI
+   * @throws CertFSException when cannot construct https URI
+   */
+  private URI createHttpsUri(final URI certFsUri, final boolean excludesQuery) throws CertFSException{
+    String certFsScheme = certFsUri.getScheme();
+    if (!CertFSConstants.CERTFS_URI_SCHEME.equalsIgnoreCase(certFsScheme)) {
+      throw new CertFSException("Invalid scheme for certFsUri: " + certFsUri);
+    }
+    try {
+      String query = excludesQuery ? null : certFsUri.getQuery();
+      URI httpsUri = new URI(
+          CertFSConstants.HTTPS_URI_SCHEME,
+          certFsUri.getUserInfo(),
+          certFsUri.getHost(),
+          certFsUri.getPort(),
+          certFsUri.getPath(),
+          query,
+          certFsUri.getFragment()
+      );
+      return httpsUri;
+
+    } catch (URISyntaxException e) {
+      throw new CertFSException("cannot create the corresponding https URI due to syntax error for " + certFsUri, e);
+    }
+  }
+
+  /**
+   * get bcstyle x500name from config and replace the placeholder (such as __HOSTNAME__) if needed
+   * @return bcstyle X500Name in the format of string, e.g. "CN=myhost.com, OU=Security, O=LinkedIn, L=Mountain View, ST=California, C=US"
+   * @throws CertFSException when no valid bcstyle X500Name
+   * @throws IOException when generating hostname for bcstyle x500Name
+   */
+  private String getBcstyleX500Name() throws CertFSException, IOException {
+    // sample of raw bcstyle x500name
+    // "CN=__HOSTNAME__, OU=Security, O=LinkedIn, L=Mountain View, ST=California, C=US"
+    String bcstyleX500NameRaw = getConf().get(CertFSConstants.BCSTYLE_X500NAME);
+    if (null == bcstyleX500NameRaw) {
+      throw new CertFSException("no csr.ssl.bcstyle.x500name configured");
+    }
+    String bcstyleX500Name;
+    if (bcstyleX500NameRaw.indexOf(CertFSConstants.HOSTNAME_PLACEHOLDER) >=0) {//contains placeholder
+      bcstyleX500Name = bcstyleX500NameRaw.replace(CertFSConstants.HOSTNAME_PLACEHOLDER, getHostName());
+    } else {
+      bcstyleX500Name = bcstyleX500NameRaw;
+    }
+    return bcstyleX500Name;
+  }
+
+  /**
+   * create the certificate signing request with the giving {@link BCStyle} {@link X500Name}
+   * @param bcstyleX500Name in String format, e.g."CN=hostname.com, OU=Security, O=LinkedIn, L=Mountain View, ST=California, C=US"
+   * @return Certificate signing request in string format
+   * @throws CertFSException certificate local signing issue
+   * @throws IOException reader/writer related exception
+   */
+  private String createCSR(final String bcstyleX500Name) throws CertFSException, IOException {
```

Comment:
This is a standard approach to create CSR, but we will leave it up to the user to define the sign algorithm in the config and generate x500 name in the interface implementation.